### ISSUE TITLE
Time with a timezone is parsed into type Parsed

### ISSUE THREAD

**Issue by baldimir** [apache/incubator-kie-issues#69]
Time strings containing timezone, like `@"10:10:10@Australia/Melbourne"` are parsed into a Java type `Parsed` in Drools DMN engine. This type is not supported in the operations in the InfixOpNode, so when it is used in e.g. addition or subtraction, the engine returns null for cases when a result should be returned. 

Investigate and fix this behaviour. Find out if parsing these time strings into `Parsed` type is an accident (uncovered some condition when parsing) or if we should support it in `InfixOpNode` and fix the problem based on the investigation. 

Acceptance criteria: 
- Use cases for addition and subtraction where time with a timezone is used work. 
- New code and these use cases have 100% code coverage.

**Comment by baldimir** [comment]
Mentioning @tarilabs so he is aware of this.

**Comment by gitgabrio** [comment]
@baldimir @yesamer @kbowers-ibm 
I've digged the code a little bit.
I think the crucial point is [here](https://github.com/kiegroup/drools/blob/d2b751c25c9d6e5e749177c8d92f1cdfd494d2aa/kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/TimeFunction.java#L68)

```java
TemporalAccessor parsed = FEEL_TIME.parse(val);

            if (parsed.query(TemporalQueries.offset()) != null) {
                // it is an offset-zoned time, so I can know for certain an OffsetTime
                OffsetTime asOffSetTime = parsed.query(OffsetTime::from);
                return FEELFnResult.ofResult(asOffSetTime);
            } else if (parsed.query(TemporalQueries.zone()) == null) {
                // if it does not contain any zone information at all, then I know for certain is a local time.
                LocalTime asLocalTime = parsed.query(LocalTime::from);
                return FEELFnResult.ofResult(asLocalTime);
            }
```

When I send the string `"@\"13:20:00@Europe/Rome\""`, then ->  `parsed.query(TemporalQueries.zone()) != null `,  the `parsed` object is not transformed, and returned as it is.
There would need another `if` to deal with the above case and return a sort of "Zoned" time, but it does not exists (IINW). THe issue is that "zoned" times are based on offset (i.e. difference between UTC and timezone) but for the same zoneid, such offset maybe different depending on actual date and geographical regions.
This [document](https://docs.oracle.com/javase/tutorial/datetime/iso/timezones.html) describe that and my understanding is that, basically, a "zoned" representation of a "time-only" (i.e. hour-min-sec)  is not supported/meaningfull. I begun to write some code, but I would have to use some workarounds/assumptions that may result in wrong values.
It is possible that I am missing something, of course, but it would be worth to understand from DMN spec if that kind of input is valid in first place, so I'll look there....

**Comment by gitgabrio** [comment]
@baldimir @yesamer @kbowers-ibm
Looking [here](https://docs.camunda.io/docs/components/modeler/feel/language-guide/feel-data-types/#time) , FEEL spec cleary stated that `HH:mm:ss@ZoneId` (e.g. `@\"13:20:00@Europe/Rome\"`) has to be translated to `java.time.OffsetTime`.
Now, I'm trying to see how this could be done...

**Comment by baldimir** [comment]
That is not the spec. Those are Camunda docs. The spec is here https://www.omg.org/spec/DMN/1.4/Beta1/PDF. Be aware, Camunda may do things a bit differently. Based on TCK they don't pass all the tests.

**Comment by gitgabrio** [comment]
> That is not the spec. Those are Camunda docs. The spec is here https://www.omg.org/spec/DMN/1.4/Beta1/PDF. Be aware, Camunda may do things a bit differently. Based on TCK they don't pass all the tests.

Yes, you are right. I looked in DMN spec you mentioned [here]( https://www.omg.org/spec/DMN/1.4/Beta1/PDF)  "10.3.2.3.4 time", and found that:

`A time in the semantic domain is a value of the XML Schema time datatype. It can be represented by a sequence of
numbers for the hour, minute, second, and an optional time offset from Universal Coordinated Time (UTC). `

So, my understanding is that this -> `11:45:30+02:00` is a valid input, because it has offset (i.e. delta with UTC), but this -> `10:31:10@Europe/Paris` is not, because it uses ZoneID (which is not an Offset). 
That would be consistent with my previous observation, but anyway this stuff is tricky

**Comment by tarilabs** [comment]
> Attempt at converting "time@ZoneId" to Offsettime.

@baldimir 

FWIW I don't believe the related PR suggestion is what you really want. e.g.: at least to last time I was involved with, a IANA timezone is _not_ to be converted to an offset tz, DMN spec semantic related.
You might have more up2date info tho; I just thought about giving an headsup.

**Comment by baldimir** [comment]
Yes, this issue was about investigation if we need to convert the Parsed or not, because it doesn't work in the arithmetic operations. The finding is that we cannot do that, because a dot accessor like those failing tests Gabriele pointed out don't work. So the better solution might be to support Parsed in the arithmetic operations.

**Comment by tarilabs** [comment]
> Yes, this issue was about investigation (...)

in the spec there should be provision ~saying _`@tz` is assumed at some place/location_

so depending on which "day" it might be an offset qty or another. I believe I kept it separate for this reason, in order to support it for other purposes, but it might not have all ops per original post.

also fyi the `@tz` was added later in some minor, so helpful to keep that in mind the semantic might be described in other places of the spec.

hope this helps

**Comment by baldimir** [comment]
Pausing this as I need to investigate other issue. I will get back to this next week. Currently I have a new FEELZonedTime class that I use instead of OffsetTime and Parsed. What is missing is finish to write tests for the new class and check if everything works, like native compilation etc.

**Comment by jomarko** [comment]
@baldimir will this have an impact for the DMN editor we have. Does it produce currently some non-supported format?

**Comment by baldimir** [comment]
This doesn't have impact on the editor. The consequence of this is that some expressions return null when they are executed, but they shouldn't.

**Comment by gitgabrio** [comment]
@baldimir IS this issue stil valid ? Related PRs seems merged 🤔

**Comment by baldimir** [comment]
As far as I know, it is. But I would need to recheck.

**Comment by yesamer** [comment]
In my understanding, the Related PRs have been closed and postponed

**Comment by yesamer** [comment]
@baldimir @gitgabrio We need to fix this to solve #1323 and  #357
